src/utils/filesystem.ts

```diff
@@ -0,0 +1,81 @@
+// get the paths of files (no directories) in the directory
+// parent: "" or "/"
+export const getFileSystemDirectoryPaths = async (
+    handle: FileSystemDirectoryHandle,
+    parent = ""
+): Promise<string[]> => {
+    const paths: Array<string> = [];
+
+    for await (const [path, fileSystemHandle] of handle.entries()) {
+        if (fileSystemHandle instanceof window.FileSystemFileHandle) {
+            paths.push(`${parent}${path}`);
+        } else {
+            paths.push(
+                ...(await getFileSystemDirectoryPaths(
+                    fileSystemHandle,
+                    `${parent}${path}/`
+                ))
+            );
+        }
+    }
+
+    return paths;
+};
+
+// create the dst directory if it doesn't exist
+// return the dst directory handle
+// paths: "/dir1/dir2" => ["dir1","dir2"]
+export const createFileSystemDirectory = async (
+    handle: FileSystemDirectoryHandle,
+    paths: string[]
+) => {
+    let cur = handle;
+    while (paths.length > 0) {
+        const path = paths.shift();
+        if (!path) {
+            break;
+        }
+        cur = await cur.getDirectoryHandle(path, { create: true });
+    }
+    return cur;
+};
+
+// save file into the dst directory
+// create the dst file if it doesn't exist by default
+// path: a/b/c.jpg
+export const saveFileToFileSystemDirectory = async (
+    handle: FileSystemDirectoryHandle,
+    stream: FileSystemWriteChunkType,
+    path: string,
+    create = true
+) => {
+    const paths = path.split("/");
+    const fileName = paths.pop();
+    if (!fileName) return;
+
+    const dir = await createFileSystemDirectory(handle, paths);
+    const file = await dir.getFileHandle(fileName, { create });
+    const writable = await file.createWritable();
+    await writable.write(stream);
+    await writable.close();
+};
+
+// verify or request the permission of the readwrite permission
+export async function verifyFileSystemRWPermission(
```

Comment:
https://web.dev/file-system-access/
> Sometimes processing the to-be-saved data takes some time after the user clicks the Save button in your app. A common gotcha is to do this work before the `showSaveFilePicker()` code has run, resulting in a SecurityError `Failed to execute 'showSaveFilePicker' on 'Window': Must be handling a user gesture to show a file picker.`. Instead, get the file handle first, and only after obtaining the file handle start processing the data.

Reply:
We can just write an empty file to obtain the read-write permission.
Or obtain the permission after fetching the first file fully.

Between the user clicking and writing file, it may take some time.

Reply:
My opinion is wrong.